internal/app/ntpd/main.go

```diff
@@ -56,77 +54,21 @@ func main() {
 		server = data.Services.NTPd.Server
 	}
 
-	log.Println("Starting ntpd")
-	n := &NTP{Server: server}
-	n.Daemon()
-}
-
-// NTP contains a server address
-// and the most recent response from a query
-type NTP struct {
-	Server   string
-	Response *ntp.Response
-}
-
-// Daemon runs the control loop for query and set time
-// We dont ever want the daemon to stop, so we only log
-// errors
-func (n *NTP) Daemon() {
-	var err error
-	rando := rand.New(rand.NewSource(time.Now().UnixNano()))
-	ticker := time.NewTicker(time.Duration(rando.Intn(MAXPOLL)+MINPOLL) * time.Second)
+	n := &ntp.NTP{Server: server}
 
-	log.Println("initial query")
-	// Do an initial hard set of time to ensure clock skew isnt too far off
-	if err = n.Query(); err != nil {
-		log.Printf("error querying %s for time, %s", n.Server, err)
-	}
-	log.Printf("%+v\n", n.Response)
-	log.Println("Current time")
-	log.Println(time.Now())
-	if err = n.SetTime(); err != nil {
-		log.Printf("failed to set time, %s", err)
-	}
-	log.Println("Updated time")
-	log.Println(time.Now())
-
-	for {
-		<-ticker.C
-		// Set some variance with how frequently we poll ntp servers
-		if err = n.Query(); err != nil {
-			log.Printf("error querying %s for time, %s", n.Server, err)
-			continue
-		}
-		log.Printf("%+v\n", n.Response)
-		log.Println("Current time")
-		log.Println(time.Now())
-		if err = n.SetTime(); err != nil {
-			log.Printf("failed to set time, %s", err)
-			continue
-		}
-		log.Println("Updated time")
-		log.Println(time.Now())
-		ticker = time.NewTicker(time.Duration(rando.Intn(MAXPOLL)+MINPOLL) * time.Second)
-	}
-}
-
-// Query polls the ntp server to get back a response
-// and saves it for later use
-func (n *NTP) Query() error {
-	resp, err := ntp.Query(n.Server)
-	if err != nil {
-		return err
-	}
-	n.Response = resp
-	return nil
-}
-
-// SetTime sets the system time based on the query response
-func (n *NTP) SetTime() error {
-	// Not sure if this is the right thing to do
-	if n.Response == nil {
-		return nil
-	}
-	timeval := syscall.NsecToTimeval(n.Response.Time.UnixNano())
-	return syscall.Settimeofday(&timeval)
+	log.Println("Starting ntpd")
+	errch := make(chan error)
```

Comment:
@andrewrynhard this pattern any better than proxyd?

Reply:
Yeah I like this a lot better 👍. This is nice because we can handle an error from either.